change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing

bdhirsh opened the pull request:
pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire @andrewor14 

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #99919
* __->__ #98031
* #98030



cc @soumith @voznesenskym @penguinwu @anijain2305 @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @Xia-Weiwen @wenzhe-nrv @jiayisunx @desertfire

bdhirsh committed:
change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing

[ghstack-poisoned]

guangy10 commented on torch/_dynamo/eval_frame.py:
```diff
@@ -600,8 +600,10 @@ def export(
-        If False, exports a graph with Python operators. Default is False.
+        aten_graph (str): Valid options include:
+          "none": export a graph with Python operations. Default is False.
+          "aten": export a graph with ATen operations.
+          "aten_pre_autograd": export a graph with pre_autograd ATen operations.
```
What's the difference between pre_autograd aten ops and non-pre_autograd ones?

ezyang commented on torch/_dynamo/eval_frame.py:
```diff
@@ -584,7 +584,7 @@ class directly; instead, use :func:`torch._export.dynamic_dim`.
     f: Callable[..., Any],
     *args,
-    aten_graph: bool = False,
+    aten_graph: str = "none",
```
I don't think I would have personally gone ahead and done a BC-breaking change here. But if we are in the business of making a BC-breaking change, I think we ought to also rename the keyword argument as well. It's weird and redundant to say `aten_graph="aten"`. How about `dialect` or `target_dialect`? Also cc @SherlockNoMad for opinions.

bdhirsh replied:
`dialect` sounds good to me (but interested in Sherlock's opinion).

I'm also concerned about BC risk - if you're worried about changing the bool, I'm happy to leave it alone (although two bools to control the "dialect" feels bad). Fixing any internal call-sites will be annoying but doable.

ezyang replied:
If you're willing to put in the effort, fixing the API earlier rather than later is good. I just remember I had specifically asked you to plumb this, and had been thinking you'd do the low effort thing.

bdhirsh replied:
you're right - in interesting of landing sooner, I'm avoiding the API concern for now and adding an extra boolean argument. cc @andrewor14

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh commented:
updated, fixed tests

ezyang commented on torch/fx/experimental/proxy_tensor.py:
```diff
@@ -414,7 +414,8 @@ def can_handle_tensor(x):
         constant = None
 
-    track_tensor_tree(out, proxy_out, constant=constant, tracer=tracer)
+    with inside_mode(proxy_mode):
```
This seems surprising to me.

ezyang replied:
still waiting on explainer

bdhirsh replied:
sigh, wrote the comment when I initially made this update and forgot to submit it.

I'd be happy to make this a separate PR. The issue was that:

(1) We detach tensors as part of creating proxies.

(2) In pre_autograd tracing, we currently push `TorchProxyDispatchMode` onto **both** the autograd mode stack, and the original python key mode stack

(3) The detach() calls get intercepted by the proxy mode on the original python key mode stack.

So we need to be careful that any aten ops that we call inside of TorchProxyDispatchMode happen in a `with inside_mode()` now.

I think we agreed 2-3 weeks ago that we could avoid re-entrant issues if we solved the "fallthrough keys can't be intercepted by the python dispatcher" issue, but we agreed this is difficult.

ezyang replied:
Ok put this comment in the code?

ezyang commented:
This looks fine EXCEPT for the one line at the end, how come?

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh committed:
Update on "change torch._dynamo.export(aten_graph=...) to allow pre_autograd tracing"

pre_autograd tracing is still early, but it should work for basic cases. This PR changes the API a bit for export to expose pre_autograd tracing. Name bikeshedding is welcome, but it looks like:
```
torch._dynamo.export(..., aten_graph="aten_pre_autograd")
```

cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire andrewor14 




cc soumith voznesenskym penguinwu anijain2305 EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng Xia-Weiwen wenzhe-nrv jiayisunx desertfire

[ghstack-poisoned]

bdhirsh commented:
@pytorchbot merge
